### PR TITLE
add e2e test for AKS node pool taints

### LINKED ISSUE #2860: AzureManagedCluster E2E: node pool Taints CRUD tests
/kind feature
/area managedclusters

**Describe the solution you'd like**
Create new AzureManagedCluster E2E: node pool Taints CRUD tests.

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**: This change adds a new e2e test for AzureManagedMachinePools which updates `spec.taints` and ensures those changes are reflected in AKS.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2860

**Special notes for your reviewer**:
Currently there's [no way to delete all taints from a node pool](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2682#issuecomment-1267701356), so this test has a side effect of leaving a `PreferNoSchedule` taint on each node pool that originally had no taints. I tried defining a taint with a bogus `effect` like `NoOp` but that gets rejected as invalid by AKS.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/aks_node_taints.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+				},
+				{
+					Effect: infrav1exp.TaintEffect(corev1.TaintEffectPreferNoSchedule),
+					Key:    "capz-e2e-2",
+					Value:  "test2",
+				},
+			}
+			Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(ammp), ammp)).To(Succeed())
+			ammp.Spec.Taints = expectedTaints
+			Expect(mgmtClient.Update(ctx, ammp)).To(Succeed())
```
Comment (jackfrancis) [inline test/e2e/aks_node_taints.go:121]: pending thoughts on https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3058 let's wrap the Get and Update methods inside a common retry loop

Comment (nojnhuh) [reply test/e2e/aks_node_taints.go]: done

File: test/e2e/aks_node_taints.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+func AKSNodeTaintsSpec(ctx context.Context, inputGetter func() AKSNodeTaintsSpecInput) {
+	input := inputGetter()
+
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	auth, err := settings.GetAuthorizer()
+	Expect(err).NotTo(HaveOccurred())
+
+	managedclustersClient := containerservice.NewManagedClustersClient(subscriptionID)
```
Comment (jackfrancis) [inline test/e2e/aks_node_taints.go:54]: Another one where it seems we're not actually doing anything w/ this client object.

Btw these test PRs are awesome!

File: test/e2e/aks_node_taints.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		go func(mp *expv1.MachinePool) {
+			defer GinkgoRecover()
+			defer wg.Done()
+
+			ammp := &infrav1.AzureManagedMachinePool{}
+			Expect(mgmtClient.Get(ctx, types.NamespacedName{
+				Namespace: mp.Spec.Template.Spec.InfrastructureRef.Namespace,
+				Name:      mp.Spec.Template.Spec.InfrastructureRef.Name,
+			}, ammp)).To(Succeed())
+			initialTaints := ammp.Spec.Taints
```
Comment (jackfrancis) [inline test/e2e/aks_node_taints.go:80]: So this setting (probably should have asked this on previous E2E tests that followed this pattern) will keep our reconcile code honest in terms of being able to handle the various combinatoric values that could be here:

- actual taints
- empty set
- zero value

I know we aren't going to have the cluster template coverage to test all of the above, but just wanna be clear that if we *do* mutate the templates going forward and in fact there is an edge case in terms of our reconciler not doing the right thing from `nil`--> some value or some other such thing that we've stubbed our toe on in the past, this E2E test will fail, right? (fail == what we want in this case :))

Comment (nojnhuh) [reply test/e2e/aks_node_taints.go:80]: If the template starts with a value, then this particular test will not do a nil -> non-nil update. The main hurdle with adding that step is that the from the initial non-nil to the nil state in CAPZ means "don't change the value in AKS." That might be a bit tricky to express robustly in a test since that would involve checking periodically for \*some TBD period of time\* that nothing happens https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2838#discussion_r1064898187.

Though perhaps we could still perform the update to nil, but skip checking the value in AKS normally following that. From there we can verify a nil -> non-nil update like we do now.

That still leaves the gap of verifying what happens after a non-nil -> nil update, but that's also not being tested currently and should be simple to slot in as necessary once we iron out that particular logic.

Comment (jackfrancis) [reply test/e2e/aks_node_taints.go:80]: ACK
